compiler/stable_mir/src/mir/alloc.rs

```diff
@@ -47,13 +48,39 @@ impl GlobalAlloc {
 #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
 pub struct AllocId(usize);
 
+fn get_index_and_populate_allocs(scc: &mut SerializeCycleCheck, alloc_id: &AllocId) -> usize {
+    let galloc = &GlobalAlloc::from(*alloc_id);
+    if !scc.seen_allocs.contains(alloc_id) {
+        scc.seen_allocs.insert(*alloc_id);
+        scc.gallocs_ordered.push(galloc.clone());
+        match galloc {
+            GlobalAlloc::Memory(allocation) => {
+                (&allocation.provenance.ptrs)
+                    .into_iter()
+                    .for_each(|(_, prov)| { get_index_and_populate_allocs(scc, &prov.0); })
+            },
+            _ => {},
+        }
+        scc.seen_allocs.len() - 1
+    } else {
+        (&scc.gallocs_ordered)
+            .into_iter()
+            .position(|needle| galloc == needle)
+            .unwrap()
+    }
+
+}
```

Comment:
Maybe not the best way to handle things, I feel like there is a lot of cloning especially since the whole vector gets cloned in `global_allocs` function. What do you think @sskeirik?

Reply:
For now, let's not worry about this. We can always fix it in a later version.